File: medium/_1282_GroupthePeopleGiventheGroupSize.py
# 1282. Group the People Given the Group Size They Belong To

"""https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/discuss/446370/Python-HashMap"""

class Solution(object):
    def groupThePeople(self, groupSizes):
        """
        :type groupSizes: List[int]
        :rtype: List[List[int]]
        """
        count = collections.defaultdict(list)
        for i, size in enumerate(groupSizes):
            count[size].append(i)
        return [l[i:i + s]for s, l in count.items() for i in xrange(0, len(l), s)]

File: medium/_1282_GroupthePeopleGiventheGroupSize.cpp
// 1282. Group the People Given the Group Size They Belong To


class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
		auto max = *max_element(std::begin(groupSizes), std::end(groupSizes));
		vector<vector<int>> collects(max);
		for (int i = 0; i < groupSizes.size(); ++i)
		{
			collects[groupSizes[i] - 1].push_back(i);
		}
		
		vector<vector<int>> output;
		for (auto & collect : collects)
		{
			if (collect.size())
			{
				auto num = groupSizes[collect[0]];
				for (size_t i = 0; i < collect.size() / num; i++)
				{
					output.push_back(vector<int>(collect.begin() + i, collect.begin() + i + num));
				}
			}
		}

		return output;    
    }
};